api/grpcserver/mesh_service.go

```diff
@@ -0,0 +1,360 @@
+		destTxIDs := s.Mesh.GetTransactionsByDestination(layerID, addr)
+		txIDs = append(txIDs, destTxIDs...)
+		originTxIds := s.Mesh.GetTransactionsByOrigin(layerID, addr)
+		txIDs = append(txIDs, originTxIds...)
+	}
+	return txIDs
+}
+
+func convertTransaction(t *types.Transaction) *pb.Transaction {
+	return &pb.Transaction{
+		Id: &pb.TransactionId{Id: t.ID().Bytes()},
+		Data: &pb.Transaction_CoinTransfer{
+			CoinTransfer: &pb.CoinTransferTransaction{
+				Receiver: &pb.AccountId{Address: t.Recipient.Bytes()},
+			},
+		},
+		Sender: &pb.AccountId{Address: t.Origin().Bytes()},
+		GasOffered: &pb.GasOffered{
+			// TODO: Check this math! Does fee == price?
+			GasPrice:    t.Fee,
+			GasProvided: t.GasLimit,
+		},
+		Amount:  &pb.Amount{Value: t.Amount},
+		Counter: t.AccountNonce,
+		Signature: &pb.Signature{
+			// TODO: confirm default signature scheme
+			Scheme:    pb.Signature_SCHEME_ED25519_PLUS_PLUS,
+			Signature: t.Signature[:],
+			PublicKey: t.Origin().Bytes(),
+		},
+	}
+}
+
+func convertActivation(a *types.ActivationTx) (*pb.Activation, error) {
+	// TODO: It's suboptimal to have to serialize every atx to get its
+	// size but size is not stored as an attribute.
+	data, err := a.InnerBytes()
+	if err != nil {
+		return nil, err
+	}
+	return &pb.Activation{
+		Id:             &pb.ActivationId{Id: a.ID().Bytes()},
+		Layer:          a.PubLayerID.Uint64(),
+		SmesherId:      &pb.SmesherId{Id: a.NodeID.ToBytes()},
+		Coinbase:       &pb.AccountId{Address: a.Coinbase.Bytes()},
+		PrevAtx:        &pb.ActivationId{Id: a.PrevATXID.Bytes()},
+		CommitmentSize: uint64(len(data)),
+	}, nil
+}
+
+// LayersQuery returns all mesh data, layer by layer
+func (s MeshService) LayersQuery(ctx context.Context, in *pb.LayersQueryRequest) (*pb.LayersQueryResponse, error) {
+	log.Info("GRPC MeshService.LayersQuery")
+
+	// last validated layer
+	// TODO: Do we need to distinguish between layers that have passed tortoise,
+	// passed hare, been applied to state, etc.? For now eveything is either
+	// unspecified, or fully confirmed.
+	// see https://github.com/spacemeshos/api/issues/89
+	//lastValidLayer := s.Mesh.ProcessedLayer()
+	lastValidLayer := s.Mesh.LatestLayerInState()
+
+	layers := []*pb.Layer{}
+	for l := uint64(in.StartLayer); l <= uint64(in.EndLayer); l++ {
+		layerStatus := pb.Layer_LAYER_STATUS_UNSPECIFIED
+		if l <= lastValidLayer.Uint64() {
+			layerStatus = pb.Layer_LAYER_STATUS_CONFIRMED
+		}
+
+		layer, err := s.Mesh.GetLayer(types.LayerID(l))
+		// TODO: Be careful with how we handle missing layers here.
+		// A layer that's newer than the currentLayer (defined above)
+		// is clearly an input error. A missing layer that's older than
+		// lastValidLayer is clearly an internal error. A missing layer
+		// between these two is a gray area: do we define this as an
+		// internal or an input error? For now, all missing layers produce
+		// internal errors.
+		if layer == nil || err != nil {
```

Comment:
We need to figure out how to handle errors here. @antonlerner @avive could use your input on this. Right now, as the comment says, any time we try to get a layer that we don't have data for, it returns an "internal" error. Should we check the requested layer numbers and instead return an "invalid input" error if the caller requests a layer that's newer than the current layer? What about a layer that's <= current layer, but > our last known layer? Is that an internal or an invalid input error?

Reply:
we shouldn't return any API responses when syncing (return error) since we cannot provide the info. if an incorrent layer was requested return error to client

Reply:
@antonlerner What's your definition of "incorrect layer"?

Reply:
an error we haven't reached yet

Reply:
let's follow up on this in https://github.com/spacemeshos/api/issues/100